### ISSUE TITLE
Equihash: Fix memory ballooning problem at large k

### ISSUE THREAD

**Issue by str4d** [zcash/zcash#1165]
As `k` is increased (and thus the number of collision rounds), the chance of generating rows with duplicate indices increases (since in `OptimisedSolve()` we are truncating indices, and thus cannot eliminate duplicates). These duplicate rows result in all-zero hashes, which will get sorted to the top of the next round's list. If we end up with `j` all-zero hashes (which empirically has been in the tens to hundreds, compared to between one and ~six for regular collisions), the collision step for the next round will attempt to generate `(j-1)!` new rows, which quickly causes "bad alloc" errors if you don't have enough memory.

To fix this, I propose that we eliminate all-zero rows after the sorting step, if they are duplicated (ie. there are other all-zero rows that have the same truncated indices when sorted). This will risk eliminating a very small fraction of legitimate solutions, but that should hopefully be statistically insignificant compared to the likelihood of it being a genuine collision.

We could also go by the metric "add the truncated indices to a set, and drop the all-zero row if the set is half the expected size" (ie. every truncated index was duplicated), again risking eliminating valid solutions but hopefully inconsequentially. By the time we are likely to get an all-zero row in the non-final round due to legitimate collisions, we will have enough partial indices in each row that it should be statistically unlikely that all of them collide on their first 8 bits.

**Comment by str4d** [comment]
I don't think we should just eliminate any all-zero rows that occur in a non-final round. While it is statistically unlikely that this should occur, it is *more* likely than either of the above two cases.

**Comment by daira** [comment]
Let's also have a failsafe that aborts a solution run if the number of solutions goes more than 50% above the expected number, to avoid being obnoxious about memory usage.

**Comment by str4d** [comment]
I've actually had a change of mind: I now thing we should eliminate partials in a final round that do in fact satisfy the `IsProbablyDuplicate` check. I've trialled locally, and it does seem that most of the partials remaining in the final round are in fact duplicates (in that every truncated index appears an even number of times).

This *will* affect the probability of finding a solution, but solutions dropped by this would need to have `2^(k-1)` pairs of indices that collide on their first 8 bits, which becomes increasingly unlikely as `k` increases. I haven't done the calculation, but I'd expect it to be far less likely than the 2 in 10,000 bug that we fixed in #1048.

If it becomes a significant issue, we can add logic that disables the dropping for small `k`.
